[wed] save wake-up parent when receiving a wake-up frame

edmont opened the pull request:
[wed] save wake-up parent when receiving a wake-up frame

This commit introduces a basic `CslNeighborTable` which currently
assumes a single enhanced CSL capable neighbor: the Wake-up Parent.

When a valid wake-up frame is received the sender's information is
used to add such `CslNeighbor` to the table. `IndirectSender` and
`CslTxScheduler` modules are refactored to be able to enqueue
messages for enhanced CSL neighbors.

Actual link establishment actions are to be added in later commits.

edmont commented:
I created this as draft PR for now since there are build conflicts between `OPENTHREAD_FTD` and `OPENTHREAD_CONFIG_WAKEUP_END_DEVICE_ENABLE`.

For the prototype the `EnhCslSender` was created because existing `OPENTHREAD_CONFIG_MAC_CSL_TRANSMITTER_ENABLE` code was not suitable for MTD: it's mainly based on `ChildTable`, whereas WED CSL transmission must be handled towards a Sleepy Parent.

However, Thread-in-Mobile spec allows FTDs to behave as WEDs, thus the conflict.

---

I'm asking here for suggestions. Some proposals:

1. Don't support FTD WED and just disable `OPENTHREAD_FTD` + `OPENTHREAD_CONFIG_WAKEUP_END_DEVICE_ENABLE` builds and use current PR code.
2. Modify `OPENTHREAD_CONFIG_MAC_CSL_TRANSMITTER_ENABLE` with separated behavior for FTD and MTD within each function. This would probably still disallow CSL Transmitter + WED combination.
3. A bit more advanced approach than option 2, modify CSL Transmitter code to handle peers as neighbors rather than children, so that new code is reusable for WED.

We need to keep in mind possible extension of the spec to support multiple SSED-to-SSED links.

Thoughts? @abtink @bukepo @zhanglongxia @Damian-Nordic @LuDuda

Damian-Nordic commented on src/core/thread/mle.hpp:
```diff
@@ -718,6 +722,27 @@ class Mle : public InstanceLocator, private NonCopyable
 
+#if OPENTHREAD_CONFIG_WAKEUP_END_DEVICE_ENABLE
+    /**
+     * Returns whether the Thread interface is currently communicating to a Wake-up Coordinator.
```
I think in the context of MLE, we might use "WakeupParent" from the latest spec proposal to have the code better prepared for the scenario where `WC != WP`.

edmont commented:
I've decided to go with option 1 (only support WED for MTD), which adheres more to the original prototype implementation. FTD support can be added later if needed.

abtink commented:
> In any case, given that the current proposal has been thoroughly tested, would you consider accepting it for now? Since public APIs are not modified, it would be easy to refactor the internal modules at any time when needed.

Thanks @edmont. I fully understand the desire to proceed with already-tested code and that reusing `IndirectSender` and `CslTxScheduler` may require some additional work. Adding a new module, `EnhCslSender` (which duplicates some of the existing code), may seem simpler and quicker, but I feel it's better to address this now.

My worry is that if we introduce this new module with the intention of refactoring later, it might become increasingly difficult to do so as new features are built on top of it. A similar example from the the past is `RadioSpinel`. There were some undesired aspects in its initial design (such as its use of multiple `select` run-loops, its blocking nature, not using `Spinel::Encoder/Decoder`, etc.), but we went ahead with the PR and code with the intention to enhance it later, which, as you may guess, has not happened and is unlikely to happen (to be fair, it is challenging to change now).

I believe that investing the time now will ultimately lead to a cleaner and more maintainable codebase. This will make it easier to add new features and enhancements in the future. I'm happy to help with this whenever I can and as my time allows. :)

---

I have been thinking about "wakeup" and its related (enhanced) CSL use cases.

It may make sense to consider defining a new `CslNeighbor` type (to track all related information for any CSL device) along with an `EnhCslNeighborTable` to track specific entries.

One use case that was discussed is to allow a "wakeup" connection to a device already in the Thread mesh (say, acting as a sleepy child). While remaining attached, we want to allow a Wakeup Coordinator device (e.g., a phone) to use "wakeup" to connect to it. To realize this, the device would have its regular parent within the network and a separate eCSL/wakeup parent, so it would be good to track them separately.

There were also discussions about making the eCSL (duty-cycled behavior) even more symmetrical. Instead of assuming a child/parent relationship, we could say it is a link between two "eCSL neighbors."

Here's a possible class hierarchy:

- `Neighbor`: Base class for all neighbor types.
- `CslNeighbor`: Subclass of `Neighbor`, inheriting from `IndirectSender::NeighborInfo` and `CslTxScheduler::NeighborInfo` to manage CSL-specific information and behavior.
- `Child`: Subclass of `CslNeighbor`, inheriting CSL functionalities while adding extra information for its role as a child node.

This way, `IndirectSender` and `CslTxScheduler` can use `CslNeighbor` as a common type (which may be a `Child` or an entry from the new `EnhCslNeighborTable`).

`EnhCslNeighborTable` can track neighbors with which we have an eCSL link. Initially, it can be a single device, the "wakeup" parent.

I think this is similar to what you suggested as option 4 in comment above.

What do you think?

edmont commented:
Having `Child` inheriting from `CslNeighbor` seems a bit artificial, but I agree it simplifies the code for those modules.

In any case, `IndirectSender` will still have a lot of complexity since (assumption is that WED enables `OPENTHREAD_CONFIG_MAC_CSL_TRANSMITTER_ENABLE`):

- WED doesn't need `DataPollHandler`.
- In functions like `PrepareFrameForCslNeighbor` or `HandleSentFrameToCslNeighbor` you added such code:
```
#if OPENTHREAD_FTD
    // `CslNeighbor` can only be a `Child` for now, but can be changed later.
    error = PrepareFrameForChild(aFrame, aContext, static_cast<Child &>(aCslNeighbor));
#else
    OT_UNUSED_VARIABLE(aFrame);
    OT_UNUSED_VARIABLE(aContext);
    OT_UNUSED_VARIABLE(aCslNeighbor);
#endif
```
But we need to consider the case of FTD+WED. How to determine which kind of `Neighbor` we are handling? Maybe we need some kind of `GetType()` method in `Neighbor`?
- We'll need to duplicate methods that handle `ChildTable` to handle `EnhCslNeighborTable` (okay, doable).

abtink commented:
> But we need to consider the case of FTD+WED. How to determine which kind of `Neighbor` we are handling? Maybe we need some kind of `GetType()` method in `Neighbor`?

My suggestion would be to use methods like `Get<ChildTable>().Contains(const Neighbor &aNeighbor)`, which checks if a given neighbor instance is from the "child table" (if it is, we can then safely cast it to `Child`). We have a similar one for `RouterTable`. The common `PrepareFrameForCslNeighbor` or `HandleSentFrameToCslNeighbor` can then dispatch the call to the proper method (handle it as a `Child` neighbor or call a WED-specific one).

> We'll need to duplicate methods that handle `ChildTable` to handle `EnhCslNeighborTable` (okay, doable).

Yes, exactly. My suggestion is to keep the existing methods in `IndirectSender` unchanged, such as `AddMessageForSleepyChild()`, `RemoveMessageFromSleepyChild()`, `ClearAllMessagesForSleepyChild()`, etc., that deal with a sleepy `Child` (and say so in their name). We can then add new (similar) methods for new FTD/WED-related interactions in `IndirectSender`.

Within the `IndirectSender` module itself, we can refactor any common, repeated behavior or code into `private` methods that are then used by both `AddMessageForSleepyChild()` and the WED-specific ones. We can use `CslNeighbor` as a common way to refer to the neighbor.

abtink reviewed (commented):
Thanks @edmont for changing the PR.
Looks good overall.  I went through the `IndirectSender` changes. Will check the mle part later. :)

edmont commented on src/core/thread/indirect_sender.cpp:
```diff
@@ -88,7 +88,82 @@ void IndirectSender::Stop(void)
+    {
+        message = FindQueuedMessageForSleepyChild(static_cast<const Child &>(aNeighbor), AcceptAnyMessage);
+    }
+    else
```
Right, the problem is that doing so will break the non-FTD build as well. We need to be able to support:
- FTD only
- FTD + WED
- Non-FTD + WED

Similar situation happened at `IndirectSender::PrepareFrameForCslNeighbor()`, I missed fixing this one. I'll fix it here by checking for `message == nullptr` when `OPENTHREAD_CONFIG_WAKEUP_END_DEVICE_ENABLE`.

edmont commented on src/core/thread/indirect_sender.cpp:
```diff
@@ -572,6 +600,152 @@ bool IndirectSender::AcceptSupervisionMessage(const Message &aMessage)
+
+    for (Message &message : Get<MeshForwarder>().mSendQueue)
+    {
+        if (message.IsForEnhancedCslNeighbor())
```
Yes, for now my intention was to keep it as simple as possible. But I agree in the future we might want some more advanced way to marking messages.

I'd say we can wait for some specification text to be present around multiple SSED to get some inspiration about ways of marking.

abtink reviewed (commented):
Thanks, @edmont, for all the changes, and sorry for the late review. I had some time today and went through the code in this PR.

abtink commented on src/core/thread/indirect_sender.cpp:
```diff
@@ -88,7 +88,244 @@ void IndirectSender::Stop(void)
     mEnabled = false;
 }
 
+uint16_t IndirectSender::PrepareDataFrame(Mac::TxFrame &aFrame, CslNeighbor &aNeighbor, Message &aMessage)
```
To help with this, I checked out your branch and rearranged the methods in `indirect_sender` to make the `git diff` smaller. I pushed it for now in [this branch](https://github.com/abtink/openthread/commits/abtink/pr/s2s-wed-pr/) on my fork. Please feel free to check it and cherry-pick the [commit](https://github.com/openthread/openthread/commit/eb741c01238122191de7076e83218dd4bc85bf11).

edmont replied:
Thanks! Much better now.

edmont commented:
Rebased and added suggestions.

edmont commented on src/core/mac/mac.cpp:
```diff
@@ -2662,8 +2674,23 @@ Error Mac::HandleWakeupFrame(const RxFrame &aFrame)
+    parent->SetLastRxTimestamp(rvTimestampUs + kRadioHeaderShrDuration + Get<Radio>().GetCslUncertainty() * 10);
+    parent->SetEnhCslMaxTxAttempts(retryCount);
+
+    Get<Mle::Mle>().AttachToWakeupParent();
```
I'm not sure, I actually was following that direction but went back to setting CSL synchronization here in `Mac`, same as we do in `ProcessCsl()`. We are using information from wake-up frame to establish the synchronization, so I guess it makes sense to do it where we have access to that information.

edmont commented:
Rebased and limited the scope of the PR.

edmont commented on src/core/thread/mle.cpp:
```diff
@@ -5485,6 +5612,50 @@ Error Mle::RxMessage::ReadRouteTlv(RouteTlv &aRouteTlv) const
+    return mCslNeighborTable.GetFirstCslNeighbor();
+}
+
+void Mle::AttachToWakeupParent()
```
Hi, for now I've reduced the scope of this PR to just store the wake-up parent and use it in the `IndirectSender` and related modules.

I agree that your proposal of `EnhCslLinkManager` is more flexible, but it deviates a lot from the code base that we are intending to contribute. Wondering if we could merge the up to now changes and continue with your proposal as a next step. Could you maybe prepare a base structure of `Mle::EnhCslLinkManager` as a separated PR that we can later extend?

BTW, current approach is based on the idea that all the traffic is forwarded to the wake-up parent when the eCSL link is established. What would be the implications when the two links will be active at the same time?

abtink commented on src/core/thread/mle.cpp:
```diff
@@ -341,6 +343,20 @@ void Mle::SetAttachState(AttachState aState)
     return;
 }
 
+void Mle::ResetAttachCounter(void) { mAttachCounter = 0; }
```
I think we can remove these (revert back)?

abtink commented on src/core/thread/mle.hpp:
```diff
@@ -1515,6 +1571,14 @@ class Mle : public InstanceLocator, private NonCopyable
 #endif
+#if OPENTHREAD_CONFIG_WAKEUP_END_DEVICE_ENABLE
+#if OPENTHREAD_CONFIG_MLE_ATTACH_BACKOFF_ENABLE
+    TimeMilli mAttachFireTime;
```
I guess we can remove (revert) these as well?

abtink commented on src/core/thread/indirect_sender.cpp:
```diff
@@ -528,24 +579,37 @@ void IndirectSender::HandleSentFrameToChild(const Mac::TxFrame &aFrame,
     }
+#endif
+
+    return;
```
Can remove `return`?

edmont replied:
I'm afraid its needed for this build:
`OT_CMAKE_NINJA_TARGET="ot-cli-mtd"  ./script/cmake-build simulation -DOT_WAKEUP_END_DEVICE=1`

Otherwise this error is thrown:
`error: label at end of compound statement only available with ‘-std=c++2b’ or ‘-std=gnu++2b’ [-Wc++23-extensions]`

abtink replied:
Ah! I see. Thanks. Let's keep it.

abtink reviewed (commented):
Looks good. Thanks @edmont  for all teh changes.
Couple of smaller suggestion below.

edmont commented on src/core/thread/csl_tx_scheduler.hpp:
```diff
@@ -153,6 +165,10 @@ class CslTxScheduler : public InstanceLocator, private NonCopyable
             mLastRxTimestamp; ///< Radio clock time when last frame containing CSL IE was received, in microseconds.
 
+#if OPENTHREAD_CONFIG_WAKEUP_END_DEVICE_ENABLE
+        uint8_t mCslMaxTxAttempts; ///< Override for the maximum number of enhanced CSL triggered tx attempts.
```
Agree, although optimization will only apply when `OPENTHREAD_CONFIG_WAKEUP_END_DEVICE_ENABLE` is enabled.

abtink commented on src/core/thread/mle.hpp:
```diff
@@ -753,6 +754,49 @@ class Mle : public InstanceLocator, private NonCopyable
+    /**
+     * Starts the process of attaching to a Wake-up Parent, if previously configured with `AddWakeupParent`.
+     */
+    void AttachToWakeupParent();
```
```suggestion
    void AttachToWakeupParent(void);
```

abtink reviewed (approved):
Looks great. Thanks again, @edmont , for the PR and being open to all discussions and suggestions, and for making all the changes. 👍

Damian-Nordic commented on src/core/thread/indirect_sender.cpp:
```diff
@@ -582,34 +697,27 @@ Error IndirectSender::PrepareFrameForCslNeighbor(Mac::TxFrame &aFrame,
+    if (IsChild(aCslNeighbor))
+    {
+        error = PrepareFrameForChild(aFrame, aContext, static_cast<Child &>(aCslNeighbor));
+        VerifyOrExit(error != kErrorNone);
```
```suggestion
        ExitNow();
```
We don't want to then call `PrepareFrameForEnhCslNeighbor` even if `PrepareFrameForChild` fails, right (for FTD + WED case)

edmont replied:
That's right, at this point we already know that we are dealing with a child.

Damian-Nordic commented on src/core/thread/mesh_forwarder.cpp:
```diff
@@ -686,14 +686,27 @@ void MeshForwarder::SetRxOnWhenIdle(bool aRxOnWhenIdle)
     Get<Mac::Mac>().SetRxOnWhenIdle(aRxOnWhenIdle);
 
+#if OPENTHREAD_CONFIG_WAKEUP_END_DEVICE_ENABLE
+    // Data polls not allowed in enhanced CSL mode
```
Now that the enhanced CSL is not done with the sleepy parent, is blocking data polls relevant anymore? This can be cleaned up in a followup though.

edmont replied:
Yeah, to be decided, let's keep it for now.

Damian-Nordic commented on src/core/thread/mesh_forwarder_ftd.cpp:
```diff
@@ -87,14 +88,35 @@ void MeshForwarder::SendMessage(OwnedPtr<Message> aMessagePtr)
                 }
+
+#if OPENTHREAD_CONFIG_WAKEUP_END_DEVICE_ENABLE
+                if (destinedForAll)
```
followup: I guess there's a question if we should forward link and realm-local addresses to RX off when idle enh CSL neighbors.

Damian-Nordic replied:
nevermind this, I didn't realize this code is for all nodes, including sleepy ones.

edmont committed:
[wed] save wake-up parent when receiving a wake-up frame

This commit introduces a basic `CslNeighborTable` which currently
assumes a single enhanced CSL capable neighbor: the Wake-up Parent.

When a valid wake-up frame is received the sender's information is
used to add such `CslNeighbor` to the table. `IndirectSender` and
`CslTxScheduler` modules are refactored to be able to enqueue
messages for enhanced CSL neighbors.

Actual link establishment actions are to be added in later commits.

EskoDijk commented:
@edmont The current (WIP) v2 spec proposal allows different cases of wake-up:
1. WC sends wake-up frame to WED; WED checks flags & source address and establishes a P2P eCSL link to WC.
2. WC that is a Parent sends wake-up frame to one or more WED(s);  WED(s) check flags & source address and activate their link to Parent normally (either establish, re-establish, or data poll)
3. (Future/prepared-for) Third device acts as WC, wakes a WED according to 1) or 2), but is not itself the target of the link.

Are these cases relevant for this PR? (If yes, are they supported by it?)
One question is on the term "Csl" used in the code - should this be "Ecsl" or "ECsl"? Is there a need to distinguish CSL vs eCSL in the code?   Maybe not, when CSL operation can be seen as a special case of eCSL where one of the peers does not have Tx/Rx constraints.   On the other hand, having Tx constraints is something that CSL never considered (only Rx-constraints) so maybe they are different enough.

EskoDijk commented:
PS in the v2 spec the names would likely change, e.g. Wake-up Parent -> Wake-up Peer (?).   I feel we should keep the names in sync for spec/implementation.  So would we need to wait for the names to be decided in [SPEC-1355](https://threadgroup.atlassian.net/browse/SPEC-1355) before finalizing this PR?

Damian-Nordic commented:
@EskoDijk This PR was originally introducing the attachment to a sleepy parent using eCSL but given the recent spec developments, this part was reverted and for now the PR includes the initial changes to support eCSL neighbor table. The eCSL neighbor attachment has to be implemented. So does the attachment to a non-sleepy parent (has this part been already defined somewhere, including Connection IE encoding?)

radua-nxp commented on src/core/thread/mle.cpp:
```diff
@@ -5515,6 +5518,58 @@ Error Mle::RxMessage::ReadRouteTlv(RouteTlv &aRouteTlv) const
+
+    OT_ASSERT(parent != nullptr);
+
+    // TODO: Implement the logic to attach to the wakeup parent.
```
I guess there is more code to follow since I don't see the ParentRequest actually being sent. Right? Or am I missing something?

Damian-Nordic replied:
There is. In the initial spec proposal and in the prototype that we meant to contribute there would be Parent Request initiated in this function. However, there are various discussions regarding the design of this feature, so there's an ask to make this code more flexible and open for other link methods. Hence now the PR only contains the initial support for eCSL neighbor table and storing information about the wake-up peer.

radua-nxp replied:
Ok, I thought that seing @abtink response but I didn't know the history. Thanks.

EskoDijk replied:
I just want to restate here that in the latest draft v2 spec and discussions, the term "wake-up parent" is not even existing anymore. There's a "wake-up peer" defined instead. Better names can still be proposed in SPEC-1355, or objections to name changes. The reason for the name change is that in all typical/foreseen cases (from my viewpoint) where the WED connects to a peer device via the eCSL link, it can be handled using Link Request/Link Accept so that no Parent role is involved.   And in case the WED is directed to connect to its actual Parent, or find a Parent in the Thread Network, then it uses a regular Parent/Child resync or attach and doesn't use the special eCSL link.
The option to connect on an eCSL link using Parent Request/Response is now defined as a "future" method i.e. we have defined a bit in the wake-up frame to signal this capability, but we don't use the bit yet until we can clearly say why it's needed.

So the question still stands: do we want to comply to the spec regarding naming? If so, how long to wait at most until the naming discussion is settled? (calls are being set up but this can easily take a few more weeks). If no, why would we let this diverge?
